plugin/aliasing.go

```diff
@@ -0,0 +1,97 @@
+package gcpauth
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/iam/v1"
+)
+
+type aliasData struct {
+	role       *gcpRole
+	svcAccount *iam.ServiceAccount
+	instance   *compute.Instance
+}
+
+type iamAliaser func(role *gcpRole, svcAccount *iam.ServiceAccount) (alias string)
+type gceAliaser func(role *gcpRole, instance *compute.Instance) (alias string)
+
+const (
+	defaultIAMAlias = "unique_id"
+	defaultGCEAlias = "instance_id"
+)
+
+var (
+	allowedIAMAliases = map[string]iamAliaser{
+		defaultIAMAlias: getIAMSvcAccountUniqueID,
+		"":              getIAMSvcAccountUniqueID, // For backwards compatibility
+
+		"role_id": getIAMRoleID,
+	}
+	allowedGCEAliases = map[string]gceAliaser{
+		defaultGCEAlias: getGCEInstanceID,
+		"":              getGCEInstanceID, // For backwards compatibility
+
+		"role_id": getGCERoleID,
+	}
+
+	allowedIAMAliasesSlice = iamMapKeyToSlice(allowedIAMAliases)
+	allowedGCEAliasesSlice = gceMapKeyToSlice(allowedGCEAliases)
+)
+
+func iamMapKeyToSlice(m map[string]iamAliaser) (s []string) {
```

Comment:
These could probably be reduced into a single function that accepts an interface and is type switched. Same with many of the `get..` functions below.

Reply:
Unfortunately it wouldn't really help here. The function would have the same code in it (including both versions of the for loop), but wrapped up into a single function rather than two. There isn't a generic `map` type that can be used as a variable type that would allow the type coercion needed here. Additionally it's generally frowned upon to pass the empty interface around unless necessary.

Reply:
I have generally found the same thing as well - when I have tried that approach it read awkwardly, but just in a different way. I don't think there's any "perfect" answer here, but I do like the more direct approach taken here because type conversions can be expensive.